Stream Settings: Fixing error messages

Abhishek-Balaji opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->
https://github.com/zulip/zulip/issues/11669

UPDATE:- Implemented all changes.

Solves parts 1,3,4. Part 2 (typeahead) has already been completed somewhere before.

Related discussions (for choosing message for user not found):-
https://chat.zulip.org/#narrow/stream/9-issues/topic/Unable.20to.20find.20user

PART 1 (User not found):-
![not_found](https://user-images.githubusercontent.com/42106909/77825091-1c456b80-712d-11ea-8bb5-f8de76452e53.gif)

PART 2 (No matching subs):-
![Screenshot from 2020-03-28 19-27-32](https://user-images.githubusercontent.com/42106909/77825101-29faf100-712d-11ea-80c8-ebdb3ad4d231.png)

PART 3 (No users to add):-
![add_user](https://user-images.githubusercontent.com/42106909/77825069-fddf7000-712c-11ea-807e-ad6a4dddf4e9.gif)

User email is missing
![Screenshot from 2020-04-26 20-52-44](https://user-images.githubusercontent.com/42106909/80312976-e2d44e80-8805-11ea-8f3f-869e4dc83e75.png)

Abhishek-Balaji commented:
@rishig @timabbott Could you please review this?

whoodes commented on static/js/stream_edit.js:
```diff
@@ -552,6 +552,10 @@ exports.initialize = function () {
             const error = JSON.parse(xhr.responseText);
+            // The message has beem changed to "Unable to find user" but not in the API
+            if (error.msg.slice(0, 51) === "User not authorized to execute queries on behalf of") {
+                error.msg = "Unable to find user " + error.msg.slice(51);
```
This doesn't look correct.  This failure response is generated via `zerver/views/streams.py`, and [this](https://github.com/gnprice/zulip/commit/a0e53d25aba13a3723566ccd3ce99ea8c751dbb8) commit gives some context as to the background.  So it seems we simply don't want to show this particular error message in this case.  Which is as simple as:
```js
stream_subscription_info_elem.text(i18n.t("Unable to find user XXX"));
```
on a failed invite, just grab whatever user was being searched for to replace `XXX`.

whoodes commented on static/js/stream_edit.js:
```diff
@@ -219,6 +237,16 @@ function show_subscription_settings(sub_row) {
+    // Stream subscribers' emails
+    const filtered_sub_emails = exports.get_email_of_active_subscribers(sub.subscribers);
+    // Displaying "No users to add"
+    if (JSON.stringify(active_realm_emails) === JSON.stringify(filtered_sub_emails)) {
```
It seems like there should be a more efficient way to do this, as it seems like a lot of extra work being done.  Not really sure how off the top of my head, but it seems like `if realm_members - subscribers == 0` then there's nobody else to add.

Abhishek-Balaji replied:
I cannot find a solution along this line. The solutions I found either suggest stringify or involve looping over both the arrays. Could you please help me in this a bit more?

whoodes replied:
I believe `people.get_realm_count()`  will get you the number of active users in the realm.  And `static/js/stream_data.js` has various functions that could prove useful.

Abhishek-Balaji replied:
Sorry, but I'm confused here. Are we discussing about the stringify part or the lists?
`people.get_realm count()` return size of active users. `active_realm_emails` is the list of active users (list of emails of people used in `get_realm_count()`). `filtered_sub_emails` returns only people subscribers (non-bots). I am not sure how the no. of users will help us here.

Abhishek-Balaji replied:
Do you want me to compare the list of people objects directly instead of emails?

whoodes replied:
> Sorry, but I'm confused here. Are we discussing about the stringify part or the lists?

Both, I'm curious why you chose to take this approach instead of something like
```js
if (people.get_realm_count() - stream_data.get_subscriber_count(stream_name) === 0) {
    // No more users to add
}
```

The above  and `if (JSON.stringify(active_realm_emails) === JSON.stringify(filtered_sub_emails))` are working to achieve the same goal, no?

> Do you want me to compare the list of people objects directly instead of emails?

My questions is what is the reasoning for aggregating a list of emails and then comparing string representations of those lists as a means to determine if there are no more users to add?  

These are important things to put in a commit message and/or PR description, Ie, explaining why you made the changes, and why they are the correct changes. :)

Abhishek-Balaji replied:
I realized the mistake I was doing. Unfortunately, this did not strike me for a long time. Thanks for pointing this out.
Compiling a list of people/emails is completely unnecessary. I'll proceed with this idea.

whoodes commented on static/js/stream_edit.js:
```diff
@@ -219,6 +237,16 @@ function show_subscription_settings(sub_row) {
         },
     });
+    // To check if all realm users are added
+    // Realm members' emails
```
Also most of these comments are unnecessary, and usually redundant of what can be explained concisely in a commit message :)

whoodes commented on static/js/stream_edit.js:
```diff
@@ -552,8 +552,15 @@ exports.initialize = function () {
-            stream_subscription_info_elem.text(error.msg)
-                .addClass("text-error").removeClass("text-success");
+            // The message has been changed to "Unable to find user" but not in the API
+            if (error.msg.slice(0, 51) === "User not authorized to execute queries on behalf of") {
```
I just noticed [this commit](https://github.com/zulip/zulip/commit/7d6ee2e10aec95b2978306d8f1acb6f7c4244be7).  It looks like we already had this behavior, and then switched to displaying the backend error.  It's probably worth it to bring this up in chat, and inquire how to proceed with this one.

Abhishek-Balaji commented:
One of the tests are failing (no idea why). I'll fix it soon.

whoodes commented:
Hey @Abhishek-Balaji, so from the discussion on chat, did you see what was agreed on moving forward?  I think from Steve's proposed wording, the error message can be along the lines: 

`Email {principal} is not recognized`.

Also, you might try taking Greg's advice:
> Oh, also, I think that function no longer has any reason to exist. The caller should just call get_active_user directly.

> In the commit that originally added it, it did something important, which was to make "this user doesn't exist" and "this user exists but is in some other realm" produce the exact same error, so as not to leak information about which users existed outside your own realm.

Basically  removing the `PrincipalError` class and `principal_to_user_profile` function (these are only used in `streams.py`, and instead just directly call `get_active_user`.

In this case you could replace all of the occurrences of `principal_to_user_profile` such as:

```python
if principals:
    people_to_unsub = set(principal_to_user_profile(
        user_profile, principal) for principal in principals)
else:    
    people_to_unsub = set([user_profile])
```

With something like:

```python
if principals:
    try:
        people_to_unsub = set(get_active_user(
            user_profile, principal) for principal in principals)
    except UserProfile.DoesNotExist:
        return json_error(_('Email {principal} is not recognized'))
else:
    people_to_unsub = set([user_profile])
``` 

If I'm understanding things correctly here.  I think this would make more sense for the `subscription: Unable to find user.` commit.  

I would also change that commit message to: `streams.py: Remove unnecessary 'PrincipalError' code.` or something along those lines. And then summarize all of this discussion surrounding this change in the commit message.

Again, look at `git log` and carefully read how those commit messages are written, it'll help you to write commits that mesh well with Zulip's standards :)

Abhishek-Balaji commented:
Thanks for the summary. There's another test case failing that is in the same file `test_subs.py` that doesn't involve this. It failed in the previous push too (this just changed the message). I'm working on it. I'll commit the changes by tonight, if it works by then.

Abhishek-Balaji commented:
![Screenshot from 2020-03-14 04-20-06](https://user-images.githubusercontent.com/42106909/76664745-66fea980-65ab-11ea-92d8-4cf7ad501746.png)
@whoodes it still shows that all checks haven't been completed on github. Do you have any idea why this happens?

whoodes commented on static/js/stream_edit.js:
```diff
@@ -219,6 +227,12 @@ function show_subscription_settings(sub_row) {
     });
+    const realm_user_count = people.get_active_human_persons().length;
+    const human_subscriber_count = exports.get_human_subscribers_count(sub.subscribers);
+    if (realm_user_count - human_subscriber_count === 0) {
```
`realm_user_count` can just be `people.get_realm_count()` then there is no need for the extra variable. Per the function definition:
```js
exports.get_realm_count = function () {
    // This returns the number of active people in our realm.  It should
    // exclude bots and deactivated users.
    return active_user_dict.size;
};
```

Abhishek-Balaji replied:
I checked this and this function includes bots too. That's why I wrote `people.get_active_human_persons().length`.
![Screenshot from 2020-03-15 18-59-45](https://user-images.githubusercontent.com/42106909/76702361-34f66000-66ef-11ea-9adb-d463dae034c3.png)

whoodes replied:
Huh, @showell is this the right behavior and the comment is just out of date?

Abhishek-Balaji replied:
@whoodes When I pointed this out in the chat, someone started a discussion and now this function is yet to be decided upon. https://github.com/zulip/zulip/pull/14216

whoodes commented on static/js/stream_edit.js:
```diff
@@ -40,6 +40,14 @@ exports.get_email_of_subscribers = function (subscribers) {
     });
 };
 
+exports.get_human_subscribers_count = function (subscribers) {
```
If there is really no pre-existing way to get the _human_ subscriber count, this function would be more appropriately placed in `stream_data.js` next to `get_subscriber_count`, and be called `get_human_subscriber_count` (just dropping the `s` in the name).

whoodes commented on zerver/views/streams.py:
```diff
@@ -261,8 +239,11 @@ def remove_subscriptions_backend(
+            people_to_unsub = set(get_active_user(
+                principal, user_profile.realm) for principal in principals)
+        except UserProfile.DoesNotExist:
+            return json_error(_('Email not recognized'), status=403)
```
These error messages should include the email (principal) in them.

whoodes commented:
> @whoodes it still shows that all checks haven't been completed on github. Do you have any idea why this happens?

This just looks like Travis being weird.

whoodes commented:
> If there is really no pre-existing way to get the human subscriber count, this function would be more appropriately placed in stream_data.js next to get_subscriber_count, and be called get_human_subscriber_count (just dropping the s in the name).

I was thinking about this a little more, and this should change should probably done as a prep commit before the `no more users` commit.

whoodes commented:
Thanks again for working on this @Abhishek-Balaji!  A few things I noticed when testing this out.
* `No users to add` is just permanently there, not sure if that's what we went.  It might be worth having that error removed dynamically in the event that a user is unsubscribed via the table below it.
* There's a blueslip `Error: Cannot read property 'filter' of undefined` as a result of `get_human_subscriber_count`, which means you need to think about what happens when a stream has no subscribers in it.
* The `Email '' not recognized` behavior from searching an empty string from before this change feels broken.  We should probably think about handling that better.
* Perhaps more thorough testing in 7da94b01fa369f358c30d0fbf8188061d21977eb?

A simple proposed solution to solving both the `Email '' not recognized` and having two error texts in the event that there are no more users to add, *and* one tries to add a user is to this: Just 
disable the add button if the form is empty or if there are no more users.  Not sure, but might be worth looking into.

Abhishek-Balaji commented:
@whoodes Thanks for the review. I'll fix it ASAP.

showell commented on static/js/stream_edit.js:
```diff
@@ -219,6 +219,11 @@ function show_subscription_settings(sub_row) {
         },
     });
+    if (people.human_realm_count() -
+        stream_data.get_human_subscriber_count(sub.subscribers) === 0) {
```
This code feels brittle to me.  We're trying to figure out if there any unsubscribed user in our realm, because otherwise there's no point trying to add users, correct?  Well, we should just calculate that set of user_ids, because don't we want to use the same users in typeahead?  And then once we have that `Set` of user_ids, we can use `.size` on the set here.

showell replied:
Perhaps a better helper function here would be in `stream_data` and would be something like `get_unsubscribed_humans_for_stream(stream_id)`.

Abhishek-Balaji replied:
Should I retain the functions I've created in the first 2 commits? As far as I've observed and discussed in the community, those functionalities don't exist.

showell replied:
I'm working on a PR to clean up the people/stream_data APIs here.

showell replied:
Once I finish that, I'll let you know and you can cherry-pick my commits into your PR and probably remove one or two of your commits.

showell replied:
See #14280.  I think you actually only need the second commit there, which will give you `stream_data.potential_subscribers()`.  Note that I think you need to be careful about which function you call this in--you want to be sure that we update the messaging any time new subscriber events come in.

I don't believe that you'll have any need for my first commit, which adds `people.get_active_human_count`, but it's obviously relevant to some conversations we've had related to this PR.

Abhishek-Balaji replied:
Thanks. I'll take a look.

Abhishek-Balaji commented:
@showell @whoodes Please take a look whenever you can

timabbott commented:
I merged #14280, so you can rebase now.

Abhishek-Balaji commented:
@timabbott rebased it.

Abhishek-Balaji commented:
A final summary of the work done:

- If no users are left to add (including bots), we display a no users to add message.
- Changed the no more subscribers message
- Changed the Email not recognized message and refactored the code (there were some redundant functions)

timabbott commented on static/js/stream_edit.js:
```diff
@@ -621,12 +632,10 @@ exports.initialize = function () {
     $("#subscriptions_table").on("click", ".sub_unsub_button", function (e) {
         const sub = get_sub_for_target(e.target);
         const stream_row = $(this).parent();
-        subs.sub_or_unsub(sub);
```
Is removing this line intentional?

timabbott commented on static/js/stream_edit.js:
```diff
@@ -621,12 +632,10 @@ exports.initialize = function () {
             exports.open_edit_panel_for_row(stream_row);
         }
         stream_ui_updates.update_regular_sub_settings(sub);
-
```
Can you undo this random whitespace change?

timabbott commented on static/js/stream_edit.js:
```diff
@@ -81,9 +81,16 @@ function get_sub_for_target(target) {
     return sub;
 }
 
+function no_users_to_add(sub) {
```
This would be much better named `update_no_users_display`.  But I'm also not sure this is a great way to do it; couldn't we use the "message on empty" feature of the `list_render` widget instead?

timabbott commented on zerver/tests/test_subs.py:
```diff
@@ -1174,7 +1174,7 @@ def test_remove_invalid_user(self) -> None:
         self.assert_json_error(
             result,
-            "User not authorized to execute queries on behalf of 'baduser@zulip.com'",
+            "Email 'baduser@zulip.com' not recognized",
```
Can you split the backend changes as a separate commit?  I'm not sure they're actually coupled to the frontend changes.

timabbott commented:
@Abhishek-Balaji thanks for working on this!  I posted a few comments.  My high-level feedback if that you should spend more time on rewriting these commits in a way that's clear, readable, doesn't have unnecessary/random changes, and doesn't have bugs (or unexplained apparent bugs).

timabbott commented:
Closing as the issue was fixed via other PRs.
